File: php/bob/bob.php
<?php

function shouting($utterance) {
  return preg_match("/[A-Z]/", $utterance) === 1 &&
         !(preg_match("/[a-z]/", $utterance) === 1);
}

function asking($utterance) {
  return preg_match("/\?\s*$/", $utterance) === 1;
}

function notSayingAnything($utterance) {
  return preg_match("#^\s*$#i", $utterance) === 1;
}

class Bob {
  public function respondTo($utterance) {
    if (shouting($utterance)) {
      return "Whoa, chill out!";
    } else if (asking($utterance)) {
      return "Sure.";
    } else if (notSayingAnything($utterance)) {
      return "Fine. Be that way!";
    } else {
      return "Whatever.";
    }
  }
}

File: lua/bob/bob.lua
local bob = {}

local function shouting(utterance)
  return utterance:match('[A-Z]') and not utterance:match('[a-z]')
end

local function asking(utterance)
  return utterance:match('?$')
end

local function not_saying_anything(utterance)
  return utterance:match('^%s*$')
end

function bob.hey(utterance)
  if shouting(utterance) then return 'Whoa, chill out!'
  elseif asking(utterance) then return 'Sure'
  elseif not_saying_anything(utterance) then return 'Fine, be that way.'
  else return 'Whatever'
  end
end

return bob
